### PR TITLE
Fix highlighting of multiple operators

### LINKED ISSUE #61646: `%` uses NodePath color even when not part of NodePath
### Godot version

1baee21

### System information

Windows 10 x64

### Issue description

Minor regression from #61440 (I assume at least)
image
It shouldn't be green.

### Steps to reproduce

1. Use `%` as modulo operator
2. Notice it's colored like NodePath instead of operator

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by MewPurPur**
`&`, `^`, `%` will be colored like symbols when they are binary (i.e. not used as literals.)

Sequences of `&` or `^` will be colored like symbols, fixing the highlighting of `&&`, except for when a StringName or NodePath is expected.

Unary mathematical operators – `+`, `-`, and the bitwise `~` – will be colored like numbers.

Bonus: Used `line_length` and `is_whitespace()` throughout the highlighter for consistency.

| Before | After |
| ------------- | ------------- |
| old-unary-op | new-unary-op |

Fixes #61646
Implements most of https://github.com/godotengine/godot-proposals/issues/4874
Supersedes PR #63044 (It only had the `%` highlighting fix)

### CODE REVIEW

**Comment by MewPurPur** [comment]
Realized today this PR will put a number of highlighting bugs in the spotlight, even though I didn't actually introduce them.

image

Ligatures are colored like their last character, and we already have a problematic case with `]#` and `<$`. This PR introduces quite a few new problems. These ligatures aren't relevant to GDScript and should probably be broken up anyway. See https://github.com/godotengine/godot/issues/63605#issuecomment-1199105222

`&&` is a ligature and always binary. Currently it's highlighted pink-white, which makes the ligature highlighted in white. With the PR it's colored white-pink, which makes the ligature colored in pink.

**Comment by MewPurPur** [comment]
Figured out a good (in my opinion) way to fix the second issue without using the tokenizer.

**Comment by MewPurPur** [comment]
Added `~` to the rule since it's also a mathematical unary operator.

**Comment by MewPurPur** [comment]
Caught up highlighting to handle all proper syntax around unary/binary & and ^ without spacing. To the two people who would ever use such horrible syntax...

image

It's highlighted properly now! You're welcome.

File: modules/gdscript/editor/gdscript_highlighter.cpp [hunk 1]
```diff
@@ -395,13 +397,41 @@ Dictionary GDScriptSyntaxHighlighter::_get_line_syntax_highlighting_impl(int p_l
 			in_member_variable = false;
 		}
 
-		if (!in_node_path && in_region == -1 && (str[j] == '^')) {
+		if (j > 0 && (str[j] == '&' || str[j] == '^' || str[j] == '%' || str[j] == '+' || str[j] == '-' || str[j] == '~')) {
+			int k = j - 1;
+			while (k > 0 && (str[k] == '\t' || str[k] == ' ')) {
```
Comment (Paulb23) [inline modules/gdscript/editor/gdscript_highlighter.cpp:402]: Can use the `is_whitespace` macro instead, not sure if it's worth cleaning up the other places.

```suggestion
			while (k > 0 && is_whitespace(str[k])) {
```

File: modules/gdscript/editor/gdscript_highlighter.cpp [hunk 2]
```diff
@@ -46,6 +46,7 @@ Dictionary GDScriptSyntaxHighlighter::_get_line_syntax_highlighting_impl(int p_l
 
 	bool prev_is_char = false;
 	bool prev_is_number = false;
+	bool prev_is_binary_op = false; // Only checked when necessary
```
Comment (Paulb23) [inline modules/gdscript/editor/gdscript_highlighter.cpp:49]: Comment doesn't add anything, can remove it.

File: modules/gdscript/editor/gdscript_highlighter.cpp [hunk 3]
```diff
@@ -84,16 +85,17 @@ Dictionary GDScriptSyntaxHighlighter::_get_line_syntax_highlighting_impl(int p_l
...
 	}
-	for (int j = 0; j < str.length(); j++) {
+	for (int j = 0; j < line_length; j++) {
 		Dictionary highlighter_info;
 
 		color = font_color;
 		bool is_char = !is_symbol(str[j]);
 		bool is_a_symbol = is_symbol(str[j]);
 		bool is_number = is_digit(str[j]);
+		bool is_binary_op = false; // Only checked when necessary
```
Comment (Paulb23) [inline modules/gdscript/editor/gdscript_highlighter.cpp:98]: Same here.

**Review by Paulb23 — APPROVED** [review]
Code looks good, couple nitpicks inline, but don't think there is a clean way to handle this without moving to using to tokenizer.

File: modules/gdscript/editor/gdscript_highlighter.cpp (hunk 1 above)
Comment (MewPurPur) [reply modules/gdscript/editor/gdscript_highlighter.cpp:402]: I'll clean up other places too.

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Thank you and congrats on your first contribution to Godot Engine's source! 🎉

**Pull request merged**